sdk/python/kfp/v2/compiler/compiler.py

```diff
@@ -0,0 +1,841 @@
+# Copyright 2018-2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import datetime
+import json
+from collections import defaultdict
+from deprecated import deprecated
+import inspect
+import tarfile
+import uuid
+import zipfile
+from typing import Callable, Set, List, Text, Dict, Tuple, Any, Union, Optional
+
+import kfp
+from kfp.dsl import _for_loop
+
+from kfp import dsl
+from kfp.compiler._k8s_helper import sanitize_k8s_name
+from kfp.compiler._op_to_template import _op_to_template
+
+from kfp.components.structures import ComponentSpec, InputSpec
+from kfp.components._yaml_utils import dump_yaml
+from kfp.dsl._metadata import _extract_pipeline_metadata
+from kfp.dsl._ops_group import OpsGroup
+from kfp.dsl import ir_types
+from kfp.v2.proto import pipeline_spec_pb2
+from google.protobuf.json_format import MessageToJson
+
+
+class Compiler(object):
+  """DSL Compiler that compiles pipeline functions into workflow yaml.
+
+  Example:
+    How to use the compiler to construct workflow yaml::
+
+      @dsl.pipeline(
+        name='name',
+        description='description'
+      )
+      def my_pipeline(a: int = 1, b: str = "default value"):
+        ...
+
+      Compiler().compile(my_pipeline, 'path/to/workflow.yaml')
+  """
+
+  def _pipelineparam_full_name(self, param):
+    """_pipelineparam_full_name converts the names of pipeline parameters
+      to unique names in the argo yaml
+
+    Args:
+      param(PipelineParam): pipeline parameter
+      """
+    if param.op_name:
+      return param.op_name + '-' + param.name
+    return param.name
+
+  def _get_groups_for_ops(self, root_group):
+    """Helper function to get belonging groups for each op.
+
+    Each pipeline has a root group. Each group has a list of operators (leaf) and groups.
+    This function traverse the tree and get all ancestor groups for all operators.
+
+    Returns:
+      A dict. Key is the operator's name. Value is a list of ancestor groups including the
+              op itself. The list of a given operator is sorted in a way that the farthest
+              group is the first and operator itself is the last.
+    """
+    def _get_op_groups_helper(current_groups, ops_to_groups):
+      root_group = current_groups[-1]
+      for g in root_group.groups:
+        # Add recursive opsgroup in the ops_to_groups
```

Comment:
Do we support recursive opsgroup right now?

If not maybe a TODO and a NotImplemented error can make this more explicit?